torch_xla/experimental/custom_kernel_from_jax.py

```diff
@@ -0,0 +1,397 @@
+    )
+    return (q_grad, k_grad, v_grad)
+
+
+@custom_op("xla::sa_custom_forward", mutates_args=())
+def sa_custom_forward(
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    config: str,
+    decoder_segment_ids: torch.Tensor | None,
+    attn_logits_soft_cap: float | None,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+  return tpu_splash_attention_jax_call_wrapper(
+      q,
+      k,
+      v,
+      config,
+      decoder_segment_ids,
+      attn_logits_soft_cap,
+      is_forward=True,
+      grad_output=None,
+  )
+
+
+@sa_custom_forward.register_fake
+def sa_custom_forward_fake(
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    config: str,
+    decoder_segment_ids: torch.Tensor | None,
+    attn_logits_soft_cap: float | None,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+  # q.shape: batch_size, seq_length, num_heads, kv (head_dim?)
+  return (torch.empty_like(q), None, None)
+
+
+@custom_op("xla::sa_custom_backward", mutates_args=())
+def sa_custom_backward(
+    grad_output: torch.Tensor,
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    config: str,
+    decoder_segment_ids: torch.Tensor | None,
+    attn_logits_soft_cap: float | None,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+  o = tpu_splash_attention_jax_call_wrapper(
+      q,
+      k,
+      v,
+      config,
+      decoder_segment_ids,
+      attn_logits_soft_cap,
+      is_forward=False,
+      grad_output=grad_output,
+  )
+  return o
+
+
+@sa_custom_backward.register_fake
+def sa_custom_backward_fake(
+    grad_output: torch.Tensor,
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    config: str,
+    decoder_segment_ids: torch.Tensor | None,
+    attn_logits_soft_cap: float | None,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+  return (torch.empty_like(q), torch.empty_like(k), torch.empty_like(v))
+
+
+class SplashAttention(torch.autograd.Function):
+
+  @staticmethod
+  @requires_jax
+  def forward(ctx, q, k, v, config, decoder_segment_ids, attn_logits_soft_cap):
+    output = sa_custom_forward(q, k, v, config, decoder_segment_ids,
+                               attn_logits_soft_cap)[0]
+    ctx.save_for_backward(q, k, v, decoder_segment_ids, attn_logits_soft_cap)
+    ctx.config = config
+    return output
+
+  @staticmethod
+  @requires_jax
+  def backward(ctx, grad_output):
+    q, k, v, decoder_segment_ids, attn_logits_soft_cap = ctx.saved_tensors
+    config = ctx.config
+    grad_q, grad_k, grad_v = sa_custom_backward(grad_output, q, k, v, config,
+                                                decoder_segment_ids,
+                                                attn_logits_soft_cap)
+    return grad_q, grad_k, grad_v, None, None, None
+
+
+def splash_attention(
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    config: str,
+    decoder_segment_ids: torch.Tensor | None = None,
```

Comment:
Could we at least document the segment IDs thing and the soft cap in triple quote style doc comments? 

Maybe also document that splash attention vs flash attention, e.g. splash attention can be faster if the attention mask is sparse by skipping blocks, etc etc